shared.nix

```diff
@@ -0,0 +1,90 @@
+let sources = import ./nix/sources.nix; in
+{ haskellNix ? import sources.haskell-nix { sourcesOverride = sources; }
+, haskellNixOverlays ? (pkgsNew: pkgsOld: { haskell-nix = pkgsOld.haskell-nix // { hackageSrc = sources.hackage-nix; }; })
+, nixpkgsSrc ? haskellNix.sources.nixpkgs-2003
+, haskellCompiler ? "ghc883"
+, shellBuildInputs ? []
+, ...
+}:
+let
+  libtorchSrc = pkgs:
+    let src = pkgs.fetchFromGitHub {
+      owner = "stites";
+      repo = "pytorch-world";
+      rev = "6dc929a791918fff065bb40cbc3db8a62beb2a30";
+      sha256 = "140a2l1l1qnf7v2s1lblrr02mc0knsqpi06f25xj3qchpawbjd4c";
+    };
+    in (pkgs.callPackage "${src}/libtorch/release.nix" { });
+
+  libtorchOverlayCpu = pkgsNew: pkgsOld:
+    let libtorch = (libtorchSrc pkgsOld).libtorch_cpu; in
+    {
+      c10 = libtorch;
+      torch = libtorch;
+      torch_cpu = libtorch;
+    };
+
+  libtorchOverlayCuda92 = pkgsNew: pkgsOld:
+    let libtorch = (libtorchSrc pkgsOld).libtorch_cudatoolkit_9_2; in
+    {
+      c10 = libtorch;
+      torch = libtorch;
+      torch_cpu = libtorch;
+      torch_cuda = libtorch;
+    };
+
+  libtorchOverlayCuda102 = pkgsNew: pkgsOld:
+    let libtorch = (libtorchSrc pkgsOld).libtorch_cudatoolkit_10_2; in
+    {
+      c10 = libtorch;
+      torch = libtorch;
+      torch_cpu = libtorch;
+      torch_cuda = libtorch;
+    };
+
+  default =
+    { overlays
+    , pkgs ? (import nixpkgsSrc (haskellNix.nixpkgsArgs // { overlays = haskellNix.overlays ++ [ haskellNixOverlays ] ++ overlays; }))
+    , USE_CUDA ? false
+    , USE_GCC ? !USE_CUDA && pkgs.stdenv.hostPlatform.system == "x86_64-darwin"
+    }: if USE_CUDA && pkgs.stdenv.hostPlatform.system == "x86_64-darwin" then { } else rec {
+      inherit pkgs;
+      hsPkgs = pkgs.haskell-nix.cabalProject {
+        src = pkgs.haskell-nix.haskellLib.cleanGit { name = "hasktorch"; src = ./.; };
+        compiler-nix-name = haskellCompiler;
+        modules = [
+          ({ config, ... }: {
+            packages.libtorch-ffi.configureFlags = [
+              "--extra-lib-dirs=${pkgs.torch}/lib"
+              "--extra-include-dirs=${pkgs.torch}/include"
+              "--extra-include-dirs=${pkgs.torch}/include/torch/csrc/api/include"
+            ];
```

Comment:
I'm somewhat surprised you need these lines here.

I was under the impression that as long as you have the `libtorch-ffi.cabal` file setup correctly, you won't need to pass `extra-lib-dirs` and `extra-include-dirs` here.

(Although maybe you do actually need `--extra-include-dirs=${pkgs.torch}/include/torch/csrc/api/include`.  That looks like some sort of include-within-include...?)

Reply:
yeah, figuring these extra libs and includes out took a long, very long time. it's all the result of a lot of trail and error...

Reply:
I feel you.  We went through the exact same process for our nix stuff at work.

Reply:
I played around with this.  Here is what I found out.

Because the `libtorch-ffi.cabal` file specifies `torch` as an `extra-library`, the normal `extra-include-dirs` and `extra-lib-dirs` are specified by default, so you don't need to specify them explicitly here:

```console
$ git diff
diff --git a/shared.nix b/shared.nix
index 1d9f322f..834edd68 100644
--- a/shared.nix
+++ b/shared.nix
@@ -47,15 +47,17 @@ let
         src = pkgs.haskell-nix.haskellLib.cleanGit { name = "hasktorch"; src = ./.; };
         compiler-nix-name = haskellCompiler;
         modules = [
           ({ config, ... }: {
             packages.libtorch-ffi.configureFlags = [
-              "--extra-lib-dirs=${pkgs.torch}/lib"
-              "--extra-include-dirs=${pkgs.torch}/include"
               "--extra-include-dirs=${pkgs.torch}/include/torch/csrc/api/include"
             ];
             packages.libtorch-ffi.flags.cuda = USE_CUDA;
$ nix-build shared.nix -A defaultCpu.hsPkgs.libtorch-ffi.components.library
these derivations will be built:
  /nix/store/rnwjv7sc64rnsjncy7lwiwb0xjd608cy-libtorch-ffi-lib-libtorch-ffi-1.5.0.0.drv
building '/nix/store/rnwjv7sc64rnsjncy7lwiwb0xjd608cy-libtorch-ffi-lib-libtorch-ffi-1.5.0.0.drv'...
...
configuring
Configure flags:
--prefix=/nix/store/zih86zmz8ivgk0h7p90m061my1gr5y0q-libtorch-ffi-lib-libtorch-ffi-1.5.0.0 lib:libtorch-ffi --extra-include-dirs=/nix/store/yblksd7ajzpsbvill1bgmi2ba4mdz5fi-libtorch-1.5/include ... --extra-lib-dirs=/nix/store/yblksd7ajzpsbvill1bgmi2ba4mdz5fi-libtorch-1.5/lib ...
```

I guess technically you could set `packages.libtorch-ffi.components.library.configureFlags` instead of `packages.libtorch-ffi.configureFlags`, because you shouldn't need those configure flags for the tests, but it probably won't hurt anything either way.

Reply:
Also, as far I could tell, instead of directly setting `configureFlags`, you should be able to do something like:

```nix
packages.libtorch-ffi.components.library.includeDirs = [
  "${pkgs.torch}/include/torch/csrc/api/include"
];
```

However, this doesn't actually appear to work: https://github.com/input-output-hk/haskell.nix/issues/679